jetty-servlet/src/test/java/org/eclipse/jetty/servlet/InitServletTest.java

```diff
@@ -0,0 +1,152 @@
+    {
+        AtomicInteger initCount = new AtomicInteger();
+
+        @Override
+        public void init() throws ServletException
+        {
+            super.init();
+            try
+            {
+                //Make the initialization last a little while so
+                //other request can run
+                Thread.sleep(5000);
+            }
+            catch (InterruptedException e)
+            {
+                throw new ServletException(e);
+            }
+            //check the servlet hasn't been initialized before
+            int val = initCount.addAndGet(1);
+            assertThat(val, Matchers.equalTo(1));
+        }
+
+        @Override
+        public void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException
+        {
+            if (initCount.get() != 1)
+            {
+                resp.sendError(500, "Servlet not initialized!");
+            }
+        }
+    }
+    
+    private static class AsyncResponseListener implements Response.CompleteListener
+    {
+        private CountDownLatch resultsLatch;
+        private Integer[] results;
+        private AtomicInteger index = new AtomicInteger();
+        
+        public AsyncResponseListener(int count)
+        {
+            resultsLatch = new CountDownLatch(count);
+            results = new Integer[count];
+        }
+        
+        public void onComplete(Result result)
+        {
+            results[index.getAndAdd(1)] = result.getResponse().getStatus();
+            resultsLatch.countDown();
+        }
+        
+        public void awaitCompletion() throws InterruptedException
+        {
+            assertTrue(resultsLatch.await(60L, TimeUnit.SECONDS));
+        }
+        
+        public Integer[] getResults()
+        {
+            return results;
+        }
+    }
+
+    @Test
+    public void testServletInitialization() throws Exception
+    {
+        Server server = new Server(0);
+        ServletContextHandler context = new ServletContextHandler(server, "/");
+        server.setHandler(context);
+        //add a lazily instantiated servlet
+        context.addServlet(new ServletHolder(DemoServlet.class), "/*");
+        server.start();
+        final int port = ((NetworkConnector)server.getConnectors()[0]).getLocalPort();
+        HttpClient client = new HttpClient();
+        try
+        {
+            client.start();
+
+            //Expect 2 responses
+            final AsyncResponseListener l = new AsyncResponseListener(2);
+            
+            //req1
+            Request r1 = client.newRequest("http://localhost:" + port + "/r1");
+            r1.method(HttpMethod.GET).send(l);
+
+            //Need to give 1st request a head start before request2
+            Thread.sleep(1000);
+            
+            //req2
+            Request r2 = client.newRequest("http://localhost:" + port + "/r2");
+            r2.method(HttpMethod.GET).send(l);
+
+            l.awaitCompletion();
```

Comment:
See above, if you had the latch available in this method, you would do all asserts here, rather than in utility classes where are difficult to follow.